File: src/test/java/simplifying/method/calls/useFactoryMethod.java
package simplifying.method.calls;

import org.assertj.core.api.ThrowableAssert;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static simplifying.method.calls.Notification.createNotification;

public class useFactoryMethod {
    @Nested
    public class notification_should {
        @Nested
        public class throw_an_illegal_argument_exception {
            private void assertExceptionThrown(ThrowableAssert.ThrowingCallable throwingCallable) {
                assertThatExceptionOfType(IllegalArgumentException.class)
                        .isThrownBy(throwingCallable);
            }

            @Test
            public void when_channel_is_null() {
                assertExceptionThrown(() -> createNotification(null));
            }

            @Test
            public void when_channel_is_empty() {
                assertExceptionThrown(() -> createNotification(""));
            }

            @Test
            public void when_channel_is_not_authorized() {
                assertExceptionThrown(() -> createNotification("Unauthorized"));
            }
        }

        @Nested
        public class instantiate {
            private void assertNotificationNotNull(String channel) {
                Notification notification = createNotification(channel);
                assertThat(notification).isNotNull();
            }

            @Test
            public void when_channel_is_SMS() {
                assertNotificationNotNull("SMS");
            }

            @Test
            public void when_channel_is_EMAIL() {
                assertNotificationNotNull("EMAIL");

            }

            @Test
            public void when_channel_is_PUSH() {
                assertNotificationNotNull("PUSH");
            }
        }
    }
}

File: src/scala/src/test/scala/simplifying/method/calls/useFactoryMethod.scala
package simplifying.method.calls

import org.scalatest.funsuite.AnyFunSuite

class useFactoryMethod extends AnyFunSuite {
  test("throw an illegal argument exception when channel is empty") {
    assertThrows[IllegalArgumentException](new Notification(""))
  }

  test("throw an illegal argument exception when channel is not authorized") {
    assertThrows[IllegalArgumentException](new Notification("Unauthorized"))
  }

  test("instantiate when channel is SMS") {
    assert(new Notification("SMS").isInstanceOf[Notification])
  }

  test("instantiate when channel is EMAIL") {
    assert(new Notification("EMAIL").isInstanceOf[Notification])
  }

  test("instantiate when channel is PUSH") {
    assert(new Notification("PUSH").isInstanceOf[Notification])
  }
}
